// @forestadmin/forestadmin-client/dist/utils/ttl-cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Simplest TTL cache for multiple entries
 * Handles concurrent fetch by design
 */
class TTLCache {
    constructor(fetchMethod, ttl = 1000) {
        this.fetchMethod = fetchMethod;
        this.ttl = ttl;
        this.stateMap = new Map();
    }
    async fetch(key) {
        const now = Date.now();
        const state = this.stateMap.get(key);
        if (state && state.promise && state.expirationTimestamp && state.expirationTimestamp > now) {
            return state.promise;
        }
        const fetch = this.fetchMethod(key);
        this.stateMap.set(key, { promise: fetch, expirationTimestamp: now + this.ttl });
        fetch.catch(() => {
            // Don't cache rejected promises
            this.stateMap.delete(key);
        });
        return fetch;
    }
    clear() {
        this.stateMap.clear();
    }
    delete(key) {
        this.stateMap.delete(key);
    }
}
exports.default = TTLCache;
