// forest-express-mongoose/dist/services/has-many-getter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _forestExpress = _interopRequireDefault(require("forest-express"));
var _lodash = _interopRequireDefault(require("lodash"));
var _schema = _interopRequireDefault(require("../utils/schema"));
var _scopes = _interopRequireDefault(require("../utils/scopes"));
var _filtersParser = _interopRequireDefault(require("./filters-parser"));
var _searchBuilder = _interopRequireDefault(require("./search-builder"));
var _flattener = _interopRequireDefault(require("./flattener"));
var OBJECTID_REGEXP = /^[0-9a-fA-F]{24}$/;
var HasManyGetter = /*#__PURE__*/function () {
  function HasManyGetter(parentModel, model, opts, params, user) {
    (0, _classCallCheck2["default"])(this, HasManyGetter);
    this._parentModel = parentModel;
    this._model = model;
    this._params = params;
    this._opts = {
      Mongoose: model.base,
      connections: model.base.connections
    };
    this._user = user;
    this._searchBuilder = new _searchBuilder["default"](model, this._opts, params);
  }
  return (0, _createClass2["default"])(HasManyGetter, [{
    key: "_hasPagination",
    value: function _hasPagination() {
      return this._params.page && this._params.page.number;
    }
  }, {
    key: "_getLimit",
    value: function _getLimit() {
      if (this._hasPagination()) {
        return parseInt(this._params.page.number, 10) * this._params.page.size;
      }
      return 5;
    }
  }, {
    key: "_getSkip",
    value: function _getSkip() {
      if (this._hasPagination()) {
        return (parseInt(this._params.page.number, 10) - 1) * this._params.page.size;
      }
      return 0;
    }
  }, {
    key: "_getProjection",
    value: function _getProjection() {
      var projection = {};
      projection[_flattener["default"].unflattenFieldName(this._params.associationName)] = 1;
      projection._id = 0; // eslint-disable-line

      return projection;
    }
  }, {
    key: "_handlePopulate",
    value: function _handlePopulate(query) {
      var schema = _forestExpress["default"].Schemas.schemas[_schema["default"].getModelName(this._model)];
      _lodash["default"].each(schema.fields, function (field) {
        if (field.reference) {
          query.populate({
            path: field.field,
            strictPopulate: false
          });
        }
      });
    }
  }, {
    key: "_buildConditions",
    value: function () {
      var _buildConditions2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(recordIds) {
        var conditions, params, conditionsSearch, filtersParser, newFilters, newFiltersString;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              conditions = {
                $and: [{
                  _id: {
                    $in: recordIds
                  }
                }]
              };
              _context.next = 3;
              return (0, _scopes["default"])(this._params, this._model, this._user);
            case 3:
              params = _context.sent;
              if (!params.search) {
                _context.next = 9;
                break;
              }
              _context.next = 7;
              return this._searchBuilder.getConditions();
            case 7:
              conditionsSearch = _context.sent;
              conditions.$and.push(conditionsSearch);
            case 9:
              if (!params.filters) {
                _context.next = 20;
                break;
              }
              filtersParser = new _filtersParser["default"](this._model, params.timezone, this._opts);
              _context.next = 13;
              return filtersParser.replaceAllReferences(params.filters);
            case 13:
              newFilters = _context.sent;
              newFiltersString = JSON.stringify(newFilters);
              _context.t0 = conditions.$and;
              _context.next = 18;
              return filtersParser.perform(newFiltersString);
            case 18:
              _context.t1 = _context.sent;
              _context.t0.push.call(_context.t0, _context.t1);
            case 20:
              return _context.abrupt("return", conditions);
            case 21:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function _buildConditions(_x) {
        return _buildConditions2.apply(this, arguments);
      }
      return _buildConditions;
    }()
  }, {
    key: "_getRecordsAndRecordIds",
    value: function () {
      var _getRecordsAndRecordIds2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var id, parentRecords, splitted, childRecordIds, conditions, query, childRecords;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              id = this._params.recordId;
              if (OBJECTID_REGEXP.test(this._params.recordId)) {
                id = new this._opts.Mongoose.Types.ObjectId(id);
              }
              _context2.next = 4;
              return this._parentModel.aggregate().match({
                _id: id
              }).unwind("$".concat(_flattener["default"].unflattenFieldName(this._params.associationName))).project(this._getProjection()).exec();
            case 4:
              parentRecords = _context2.sent;
              splitted = this._params.associationName.split('@@@');
              childRecordIds = _lodash["default"].map(parentRecords, function (record) {
                return splitted.reduce(function (a, prop) {
                  return a ? a[prop] : null;
                }, record);
              });
              _context2.next = 9;
              return this._buildConditions(childRecordIds);
            case 9:
              conditions = _context2.sent;
              query = this._model.find(conditions);
              this._handlePopulate(query);
              _context2.next = 14;
              return query;
            case 14:
              childRecords = _context2.sent;
              return _context2.abrupt("return", [childRecords, childRecordIds]);
            case 16:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function _getRecordsAndRecordIds() {
        return _getRecordsAndRecordIds2.apply(this, arguments);
      }
      return _getRecordsAndRecordIds;
    }()
  }, {
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var _yield$this$_getRecor, _yield$this$_getRecor2, childRecords, childRecordIds, fieldSort, descending, recordsSorted, recordIdStrings, sortedChildRecords, fieldsSearched;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return this._getRecordsAndRecordIds();
            case 2:
              _yield$this$_getRecor = _context3.sent;
              _yield$this$_getRecor2 = (0, _slicedToArray2["default"])(_yield$this$_getRecor, 2);
              childRecords = _yield$this$_getRecor2[0];
              childRecordIds = _yield$this$_getRecor2[1];
              fieldSort = this._params.sort;
              descending = false;
              if (this._params.sort && this._params.sort[0] === '-') {
                fieldSort = this._params.sort.substring(1);
                descending = true;
              }
              if (fieldSort) {
                recordsSorted = _lodash["default"].sortBy(childRecords, function (record) {
                  return record[fieldSort];
                });
              } else {
                // NOTICE: Convert values to strings, so ObjectIds could be easily searched and compared.
                recordIdStrings = childRecordIds.map(function (recordId) {
                  return String(recordId);
                }); // NOTICE: indexOf could be improved by making a Map from record-ids to their index.
                recordsSorted = _lodash["default"].sortBy(childRecords, function (record) {
                  return recordIdStrings.indexOf(String(record._id));
                }); // eslint-disable-line
              }
              sortedChildRecords = descending ? recordsSorted.reverse() : recordsSorted;
              fieldsSearched = null;
              if (this._params.search) {
                fieldsSearched = this._searchBuilder.getFieldsSearched();
              }
              sortedChildRecords = _lodash["default"].slice(sortedChildRecords, this._getSkip(), this._getSkip() + this._getLimit());
              return _context3.abrupt("return", [sortedChildRecords, fieldsSearched]);
            case 15:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
  }, {
    key: "count",
    value: function () {
      var _count = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4() {
        var recordsAndRecordIds;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return this._getRecordsAndRecordIds();
            case 2:
              recordsAndRecordIds = _context4.sent;
              return _context4.abrupt("return", recordsAndRecordIds[0].length);
            case 4:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this);
      }));
      function count() {
        return _count.apply(this, arguments);
      }
      return count;
    }()
  }]);
}();
module.exports = HasManyGetter;
