// forest-express-mongoose/dist/services/line-stat-getter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _forestExpress = _interopRequireDefault(require("forest-express"));
var _lodash = _interopRequireDefault(require("lodash"));
var _momentTimezone = _interopRequireDefault(require("moment-timezone"));
var _schema = _interopRequireDefault(require("../utils/schema"));
var _scopes = _interopRequireDefault(require("../utils/scopes"));
var _queryBuilder = _interopRequireDefault(require("./query-builder"));
var LineStatGetter = /*#__PURE__*/function () {
  function LineStatGetter(model, params, opts, user) {
    (0, _classCallCheck2["default"])(this, LineStatGetter);
    this._model = model;
    this._params = params;
    this._opts = {
      Mongoose: this._model.base,
      connections: this._model.base.connections
    };
    this._user = user;
  }
  return (0, _createClass2["default"])(LineStatGetter, [{
    key: "_getReference",
    value: function _getReference(fieldName) {
      if (!fieldName) {
        return null;
      }
      var schema = _forestExpress["default"].Schemas.schemas[_schema["default"].getModelName(this._model)];
      var fieldNameWithoutSubField = fieldName.includes(':') ? fieldName.split(':')[0] : fieldName;
      var field = _lodash["default"].find(schema.fields, {
        field: fieldNameWithoutSubField
      });
      return field.reference ? field : null;
    }
  }, {
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var params, timezone, timezoneOffset, queryBuilder, jsonQuery, groupBy, sort, sum, query, records, momentRange, firstDate, lastDate;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return (0, _scopes["default"])(this._params, this._model, this._user);
            case 2:
              params = _context.sent;
              timezone = -parseInt((0, _momentTimezone["default"])().tz(params.timezone).format('Z'), 10);
              timezoneOffset = timezone * 60 * 60 * 1000;
              queryBuilder = new _queryBuilder["default"](this._model, params, this._opts);
              _context.next = 8;
              return queryBuilder.getQueryWithFiltersAndJoins(null);
            case 8:
              jsonQuery = _context.sent;
              groupBy = {};
              sort = {};
              if (!params.groupByFieldName) {
                _context.next = 27;
                break;
              }
              _context.t0 = params.timeRange;
              _context.next = _context.t0 === 'Day' ? 15 : _context.t0 === 'Week' ? 19 : _context.t0 === 'Year' ? 22 : 24;
              break;
            case 15:
              groupBy.year = {
                $year: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              groupBy.month = {
                $month: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              groupBy.day = {
                $dayOfMonth: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              return _context.abrupt("break", 26);
            case 19:
              groupBy.week = {
                $week: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              groupBy.year = {
                $year: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              return _context.abrupt("break", 26);
            case 22:
              groupBy.year = {
                $year: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              return _context.abrupt("break", 26);
            case 24:
              // Month
              groupBy.month = {
                $month: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
              groupBy.year = {
                $year: [{
                  $subtract: ["$".concat(params.groupByFieldName), timezoneOffset]
                }]
              };
            case 26:
              sort[params.groupByFieldName] = 1;
            case 27:
              sum = 1;
              if (params.aggregateFieldName) {
                sum = "$".concat(params.aggregateFieldName);
              }
              if (params.groupByFieldName) {
                jsonQuery.push({
                  $match: (0, _defineProperty2["default"])({}, params.groupByFieldName, {
                    $ne: null
                  })
                });
              }
              if (groupBy) {
                jsonQuery.push({
                  $group: (0, _defineProperty2["default"])((0, _defineProperty2["default"])({
                    _id: groupBy
                  }, params.groupByFieldName, {
                    $first: "$".concat(params.groupByFieldName)
                  }), "count", {
                    $sum: sum
                  })
                });
              }
              query = this._model.aggregate(jsonQuery);
              _context.next = 34;
              return query.sort(sort).project({
                values: {
                  key: '$_id',
                  value: '$count'
                }
              }).exec();
            case 34:
              records = _context.sent;
              if (records.length) {
                _context.next = 37;
                break;
              }
              return _context.abrupt("return", {
                value: []
              });
            case 37:
              momentRange = params.timeRange.toLowerCase();
              firstDate = LineStatGetter._setDate(records[0], momentRange);
              lastDate = LineStatGetter._setDate(records[records.length - 1], momentRange);
              records = records.map(function (record) {
                return {
                  label: LineStatGetter._formatLabel(record, momentRange),
                  values: record.values
                };
              });
              return _context.abrupt("return", {
                value: LineStatGetter._fillEmptyIntervals(records, momentRange, firstDate, lastDate)
              });
            case 42:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
  }], [{
    key: "_getFormat",
    value: function _getFormat(momentRange) {
      switch (momentRange) {
        case 'day':
          return 'DD/MM/YYYY';
        case 'week':
          return '[W]w-YYYY';
        case 'month':
          return 'MMM YY';
        case 'year':
          return 'YYYY';
        default:
          return null;
      }
    }
  }, {
    key: "_formatLabel",
    value: function _formatLabel(record, momentRange) {
      switch (momentRange) {
        case 'day':
          return (0, _momentTimezone["default"])().year(record._id.year).month(record._id.month - 1).startOf('month').add(record._id.day - 1, 'days').startOf(momentRange).format(LineStatGetter._getFormat(momentRange));
        case 'week':
          return (0, _momentTimezone["default"])().year(record._id.year).week(record._id.week).startOf(momentRange).format(LineStatGetter._getFormat(momentRange));
        case 'month':
          return (0, _momentTimezone["default"])().year(record._id.year).month(record._id.month - 1).startOf(momentRange).format(LineStatGetter._getFormat(momentRange));
        case 'year':
          return record._id.year.toString();
        default:
          return null;
      }
    }
  }, {
    key: "_setDate",
    value: function _setDate(record, momentRange) {
      switch (momentRange) {
        case 'day':
          return (0, _momentTimezone["default"])().year(record._id.year).month(record._id.month - 1).startOf('month').add(record._id.day - 1, 'days').startOf(momentRange);
        case 'week':
          return (0, _momentTimezone["default"])().year(record._id.year).week(record._id.week).startOf(momentRange);
        case 'month':
          return (0, _momentTimezone["default"])().year(record._id.year).month(record._id.month - 1).startOf(momentRange);
        case 'year':
          return (0, _momentTimezone["default"])().year(record._id.year).startOf(momentRange);
        default:
          return null;
      }
    }
  }, {
    key: "_fillEmptyIntervals",
    value: function _fillEmptyIntervals(records, momentRange, firstDate, lastDate) {
      var newRecords = [];
      var currentDate = firstDate;
      while (currentDate <= lastDate) {
        var currentLabel = currentDate.format(LineStatGetter._getFormat(momentRange));
        var currentRecord = _lodash["default"].find(records, {
          label: currentLabel
        });
        var value = currentRecord ? currentRecord.values.value : 0;
        newRecords.push({
          label: currentLabel,
          values: {
            value: value
          }
        });
        currentDate = currentDate.add(1, momentRange);
      }
      return newRecords;
    }
  }]);
}();
module.exports = LineStatGetter;
