// forest-express-mongoose/dist/services/pie-stat-getter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _forestExpress = require("forest-express");
var _lodash = _interopRequireDefault(require("lodash"));
var _moment = _interopRequireDefault(require("moment"));
var _schema = _interopRequireDefault(require("../utils/schema"));
var _scopes = _interopRequireDefault(require("../utils/scopes"));
var _queryBuilder = _interopRequireDefault(require("./query-builder"));
var PieStatGetter = /*#__PURE__*/function () {
  function PieStatGetter(model, params, opts, user) {
    (0, _classCallCheck2["default"])(this, PieStatGetter);
    this._model = model;
    this._params = params;
    this._opts = {
      Mongoose: this._model.base,
      connections: this._model.base.connections
    };
    this._user = user;
    this._schema = _forestExpress.Schemas.schemas[_schema["default"].getModelName(this._model)];
  }
  return (0, _createClass2["default"])(PieStatGetter, [{
    key: "_getReference",
    value: function _getReference(fieldName) {
      var fieldNameWithoutSubField = fieldName.includes(':') ? fieldName.split(':')[0] : fieldName;
      var currentField = _lodash["default"].find(this._schema.fields, {
        field: fieldNameWithoutSubField
      });
      return currentField.reference ? currentField : null;
    }
  }, {
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var params, field, queryBuilder, populateGroupByField, groupByFieldName, jsonQuery, query, sum, records;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return (0, _scopes["default"])(this._params, this._model, this._user);
            case 2:
              params = _context.sent;
              field = _lodash["default"].find(this._schema.fields, {
                field: params.groupByFieldName
              });
              queryBuilder = new _queryBuilder["default"](this._model, params, this._opts);
              populateGroupByField = this._getReference(params.groupByFieldName);
              groupByFieldName = populateGroupByField ? params.groupByFieldName.replace(':', '.') : params.groupByFieldName;
              _context.next = 9;
              return queryBuilder.getQueryWithFiltersAndJoins(null);
            case 9:
              jsonQuery = _context.sent;
              if (populateGroupByField) {
                queryBuilder.addJoinToQuery(populateGroupByField, jsonQuery);
              }
              query = this._model.aggregate(jsonQuery);
              sum = 1;
              if (params.aggregateFieldName) {
                sum = "$".concat(params.aggregateFieldName);
              }
              _context.next = 16;
              return query.group({
                _id: "$".concat(groupByFieldName),
                count: {
                  $sum: sum
                }
              }).project({
                key: '$_id',
                value: '$count',
                _id: false
              }).sort({
                value: -1
              }).exec();
            case 16:
              _context.t0 = _context.sent;
              records = {
                value: _context.t0
              };
              if (field && field.type === 'Date') {
                _lodash["default"].each(records.value, function (record) {
                  record.key = (0, _moment["default"])(record.key).format('DD/MM/YYYY HH:mm:ss');
                });
              }
              return _context.abrupt("return", records);
            case 20:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
  }]);
}();
module.exports = PieStatGetter;
