// forest-express-mongoose/dist/services/query-builder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _lodash = _interopRequireDefault(require("lodash"));
var _forestExpress = _interopRequireDefault(require("forest-express"));
var _schema = _interopRequireDefault(require("../utils/schema"));
var _orm = _interopRequireDefault(require("../utils/orm"));
var _searchBuilder = _interopRequireDefault(require("./search-builder"));
var _filtersParser = _interopRequireDefault(require("./filters-parser"));
var _projectionBuilder = _interopRequireDefault(require("./projection-builder"));
var _flattener = _interopRequireDefault(require("./flattener"));
var QueryBuilder = /*#__PURE__*/function () {
  function QueryBuilder(model, params, opts) {
    (0, _classCallCheck2["default"])(this, QueryBuilder);
    this._model = model;
    this._params = params;
    this._opts = opts;
    this._schema = _forestExpress["default"].Schemas.schemas[_schema["default"].getModelName(this._model)];
    this._searchBuilder = new _searchBuilder["default"](this._model, this._opts, this._params, this._schema.searchFields);
    this._filterParser = new _filtersParser["default"](this._model, this._params.timezone, this._opts);
    this._projectionBuilder = new _projectionBuilder["default"](this._schema);
  }
  return (0, _createClass2["default"])(QueryBuilder, [{
    key: "getFieldNamesRequested",
    value: function () {
      var _getFieldNamesRequested = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var associations, _this$_params$sort$sp, _this$_params$sort$sp2, associationFromSorting;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (!(!this._params.fields || !this._params.fields[this._model.modelName])) {
                _context.next = 2;
                break;
              }
              return _context.abrupt("return", null);
            case 2:
              if (!this._params.filters) {
                _context.next = 8;
                break;
              }
              _context.next = 5;
              return this._filterParser.getAssociations(this._params.filters);
            case 5:
              _context.t0 = _context.sent;
              _context.next = 9;
              break;
            case 8:
              _context.t0 = [];
            case 9:
              associations = _context.t0;
              if (this._params.sort && this._params.sort.includes('.')) {
                _this$_params$sort$sp = this._params.sort.split('.'), _this$_params$sort$sp2 = (0, _slicedToArray2["default"])(_this$_params$sort$sp, 1), associationFromSorting = _this$_params$sort$sp2[0];
                if (associationFromSorting[0] === '-') {
                  associationFromSorting = associationFromSorting.substring(1);
                }
                associations.push(associationFromSorting);
              }
              return _context.abrupt("return", _lodash["default"].union(this._params.fields[this._model.modelName].split(','), associations));
            case 12:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function getFieldNamesRequested() {
        return _getFieldNamesRequested.apply(this, arguments);
      }
      return getFieldNamesRequested;
    }()
  }, {
    key: "addProjection",
    value: function () {
      var _addProjection = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(jsonQuery) {
        var fieldNames, projection;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return this.getFieldNamesRequested();
            case 2:
              fieldNames = _context2.sent;
              _context2.next = 5;
              return this._projectionBuilder.getProjection(fieldNames);
            case 5:
              projection = _context2.sent;
              return _context2.abrupt("return", projection && jsonQuery.push(projection));
            case 7:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function addProjection(_x) {
        return _addProjection.apply(this, arguments);
      }
      return addProjection;
    }()
  }, {
    key: "addJoinToQuery",
    value: function addJoinToQuery(field, joinQuery) {
      if (field.reference && !field.isVirtual && !field.integration) {
        if (QueryBuilder._joinAlreadyExists(field, joinQuery)) return this;
        var referencedKey = _schema["default"].getReferenceField(field.reference);
        var subModel = _schema["default"].getReferenceModel(this._opts, field.reference);
        var unflattenedFieldName = _flattener["default"].unflattenFieldName(field.field);
        joinQuery.push({
          $lookup: {
            from: subModel.collection.name,
            localField: unflattenedFieldName,
            foreignField: referencedKey,
            as: unflattenedFieldName
          }
        });
        var fieldPath = unflattenedFieldName && this._model.schema.path(unflattenedFieldName);
        if (fieldPath && fieldPath.instance !== 'Array') {
          joinQuery.push({
            $unwind: {
              path: "$".concat(unflattenedFieldName),
              preserveNullAndEmptyArrays: true
            }
          });
        }
      }
      return this;
    }
  }, {
    key: "joinAllReferences",
    value: function () {
      var _joinAllReferences = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(jsonQuery, alreadyJoinedQuery) {
        var _this = this;
        var fieldNames, flattenReferenceNames;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return this.getFieldNamesRequested();
            case 2:
              fieldNames = _context3.sent;
              flattenReferenceNames = _flattener["default"].getFlattenedReferenceFieldsFromParams(this._model.modelName, this._params.fields);
              fieldNames = flattenReferenceNames.concat(fieldNames);
              this._schema.fields.forEach(function (field) {
                if (fieldNames && !fieldNames.includes(field.field) || QueryBuilder._joinAlreadyExists(field, alreadyJoinedQuery)) {
                  return;
                }
                _this.addJoinToQuery(field, jsonQuery);
              });
              return _context3.abrupt("return", this);
            case 7:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this);
      }));
      function joinAllReferences(_x2, _x3) {
        return _joinAllReferences.apply(this, arguments);
      }
      return joinAllReferences;
    }()
  }, {
    key: "_addFiltersToQuery",
    value: function () {
      var _addFiltersToQuery2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(jsonQuery) {
        var newFilters, newFiltersString;
        return _regenerator["default"].wrap(function _callee4$(_context4) {
          while (1) switch (_context4.prev = _context4.next) {
            case 0:
              _context4.next = 2;
              return this._filterParser.replaceAllReferences(this._params.filters);
            case 2:
              newFilters = _context4.sent;
              newFiltersString = JSON.stringify(newFilters);
              _context4.t0 = jsonQuery;
              _context4.next = 7;
              return this._filterParser.perform(newFiltersString);
            case 7:
              _context4.t1 = _context4.sent;
              _context4.t0.push.call(_context4.t0, _context4.t1);
            case 9:
            case "end":
              return _context4.stop();
          }
        }, _callee4, this);
      }));
      function _addFiltersToQuery(_x4) {
        return _addFiltersToQuery2.apply(this, arguments);
      }
      return _addFiltersToQuery;
    }()
  }, {
    key: "addSortToQuery",
    value: function addSortToQuery(jsonQuery) {
      var order = this._params.sort.startsWith('-') ? -1 : 1;
      var sortParam = order > 0 ? this._params.sort : this._params.sort.substring(1);
      if (this._params.sort.split('.').length > 1) {
        var _this$_params$sort$sp3 = this._params.sort.split('.');
        var _this$_params$sort$sp4 = (0, _slicedToArray2["default"])(_this$_params$sort$sp3, 1);
        sortParam = _this$_params$sort$sp4[0];
        var _this$_params$sort$sp5 = this._params.sort.split('.'),
          _this$_params$sort$sp6 = (0, _slicedToArray2["default"])(_this$_params$sort$sp5, 1),
          association = _this$_params$sort$sp6[0];
        this.addJoinToQuery(association, jsonQuery);
      }
      if (_flattener["default"]._isFieldFlattened(sortParam)) sortParam = _flattener["default"].unflattenFieldName(sortParam);
      jsonQuery.push({
        $sort: (0, _defineProperty2["default"])({}, sortParam, order)
      });
      return this;
    }
  }, {
    key: "addSkipAndLimitToQuery",
    value: function addSkipAndLimitToQuery(jsonQuery) {
      jsonQuery.push({
        $skip: this._getSkip()
      });
      jsonQuery.push({
        $limit: this._getLimit()
      });
      return this;
    }
  }, {
    key: "addCountToQuery",
    value: function addCountToQuery(jsonQuery) {
      if (_orm["default"].hasRequiredVersion(this._opts.Mongoose, '3.4')) {
        jsonQuery.push({
          $count: 'count'
        });
      } else {
        jsonQuery.push({
          $group: {
            _id: null,
            count: {
              $sum: 1
            }
          }
        });
      }
      return this;
    }
  }, {
    key: "_hasPagination",
    value: function _hasPagination() {
      return this._params.page && this._params.page.number;
    }
  }, {
    key: "_getLimit",
    value: function _getLimit() {
      return this._hasPagination() && this._params.page.size ? Number.parseInt(this._params.page.size, 10) : 10;
    }
  }, {
    key: "_getSkip",
    value: function _getSkip() {
      return this._hasPagination() ? (Number.parseInt(this._params.page.number, 10) - 1) * this._getLimit() : 0;
    }
  }, {
    key: "hasSmartFieldSearch",
    value: function hasSmartFieldSearch() {
      return this._searchBuilder.hasSmartFieldSearch;
    }
  }, {
    key: "getFieldsSearched",
    value: function getFieldsSearched() {
      return this._searchBuilder.getFieldsSearched();
    }
  }, {
    key: "getQueryWithFiltersAndJoins",
    value: function () {
      var _getQueryWithFiltersAndJoins = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(segment) {
        var jsonQuery;
        return _regenerator["default"].wrap(function _callee5$(_context5) {
          while (1) switch (_context5.prev = _context5.next) {
            case 0:
              jsonQuery = [];
              _context5.next = 3;
              return this.addFiltersAndJoins(jsonQuery, segment);
            case 3:
              return _context5.abrupt("return", jsonQuery);
            case 4:
            case "end":
              return _context5.stop();
          }
        }, _callee5, this);
      }));
      function getQueryWithFiltersAndJoins(_x5) {
        return _getQueryWithFiltersAndJoins.apply(this, arguments);
      }
      return getQueryWithFiltersAndJoins;
    }()
  }, {
    key: "addFiltersAndJoins",
    value: function () {
      var _addFiltersAndJoins = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(jsonQuery, segment) {
        var conditions;
        return _regenerator["default"].wrap(function _callee6$(_context6) {
          while (1) switch (_context6.prev = _context6.next) {
            case 0:
              conditions = [];
              if (!this._params.filters) {
                _context6.next = 4;
                break;
              }
              _context6.next = 4;
              return this._addFiltersToQuery(conditions);
            case 4:
              if (!this._params.search) {
                _context6.next = 7;
                break;
              }
              _context6.next = 7;
              return this._searchBuilder.getWhere(conditions);
            case 7:
              if (segment) {
                conditions.push(segment.where);
              }
              if (conditions.length) {
                jsonQuery.push({
                  $match: {
                    $and: conditions
                  }
                });
              }
              return _context6.abrupt("return", this);
            case 10:
            case "end":
              return _context6.stop();
          }
        }, _callee6, this);
      }));
      function addFiltersAndJoins(_x6, _x7) {
        return _addFiltersAndJoins.apply(this, arguments);
      }
      return addFiltersAndJoins;
    }()
  }], [{
    key: "_joinAlreadyExists",
    value: function _joinAlreadyExists(field, joinQuery) {
      return !!_lodash["default"].find(joinQuery, function (join) {
        return join && join.$lookup && join.$lookup.as === field.field;
      });
    }
  }]);
}();
module.exports = QueryBuilder;
