// forest-express-mongoose/dist/services/resource-getter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _ = require('lodash');
var Interface = require('forest-express');
var createError = require('http-errors');
var utils = require('../utils/schema');
var ResourcesGetter = require('./resources-getter');
var ResourceGetter = /*#__PURE__*/function () {
  function ResourceGetter(model, params, user) {
    (0, _classCallCheck2["default"])(this, ResourceGetter);
    this._model = model;
    this._params = params;
    this._user = user;
  }
  return (0, _createClass2["default"])(ResourceGetter, [{
    key: "_handlePopulate",
    value: function _handlePopulate(query) {
      var schema = Interface.Schemas.schemas[utils.getModelName(this._model)];
      _.each(schema.fields, function (field) {
        if (field.reference) {
          query.populate({
            path: field.field,
            strictPopulate: false
          });
        }
      });
    }
  }, {
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var record, query;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              record = null;
              _context.next = 3;
              return this._isAllowed();
            case 3:
              if (!_context.sent) {
                _context.next = 9;
                break;
              }
              query = this._model.findById(this._params.recordId);
              this._handlePopulate(query);
              _context.next = 8;
              return query.lean().exec();
            case 8:
              record = _context.sent;
            case 9:
              if (record) {
                _context.next = 11;
                break;
              }
              throw createError(404, "The ".concat(this._model.name, " #").concat(this._params.recordId, " does not exist."));
            case 11:
              return _context.abrupt("return", record);
            case 12:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
    /**
     * Check if the record is in scope for the given user.
     * We can't do that in a single request to mongo, as checking scopes requires using
     * an aggregation pipeline to perform $lookups
     */
  }, {
    key: "_isAllowed",
    value: (function () {
      var _isAllowed2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var params;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              params = {
                timezone: this._params.timezone,
                filters: JSON.stringify({
                  field: '_id',
                  operator: 'equal',
                  value: this._params.recordId
                })
              };
              _context2.next = 3;
              return new ResourcesGetter(this._model, null, params, this._user).count();
            case 3:
              _context2.t0 = _context2.sent;
              return _context2.abrupt("return", _context2.t0 === 1);
            case 5:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function _isAllowed() {
        return _isAllowed2.apply(this, arguments);
      }
      return _isAllowed;
    }())
  }]);
}();
module.exports = ResourceGetter;
