// forest-express-mongoose/dist/services/resources-getter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _lodash = _interopRequireDefault(require("lodash"));
var _forestExpress = _interopRequireDefault(require("forest-express"));
var _queryBuilder = _interopRequireDefault(require("./query-builder"));
var _schema = _interopRequireDefault(require("../utils/schema"));
var _scopes = _interopRequireDefault(require("../utils/scopes"));
var ResourcesGetter = /*#__PURE__*/function () {
  function ResourcesGetter(model, opts, params, user) {
    (0, _classCallCheck2["default"])(this, ResourcesGetter);
    this._model = model;
    this._opts = {
      Mongoose: this._model.base,
      connections: this._model.base.connections
    };
    this._params = params;
    this._user = user;
  }
  return (0, _createClass2["default"])(ResourcesGetter, [{
    key: "_getSegment",
    value: function _getSegment() {
      var _this = this;
      var schema = _forestExpress["default"].Schemas.schemas[_schema["default"].getModelName(this._model)];
      if (schema.segments && this._params.segment) {
        return _lodash["default"].find(schema.segments, function (currentSegment) {
          return currentSegment.name === _this._params.segment;
        });
      }
      return null;
    }
  }, {
    key: "_getSegmentCondition",
    value: function () {
      var _getSegmentCondition2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var segment, where;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              segment = this._getSegment();
              if (!(segment && segment.where && typeof segment.where === 'function')) {
                _context.next = 6;
                break;
              }
              _context.next = 4;
              return segment.where();
            case 4:
              where = _context.sent;
              return _context.abrupt("return", {
                where: where
              });
            case 6:
              return _context.abrupt("return", segment);
            case 7:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function _getSegmentCondition() {
        return _getSegmentCondition2.apply(this, arguments);
      }
      return _getSegmentCondition;
    }()
  }, {
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var params, fieldsSearched, segment, jsonQuery, queryBuilder, records;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return (0, _scopes["default"])(this._params, this._model, this._user);
            case 2:
              params = _context2.sent;
              fieldsSearched = null;
              _context2.next = 6;
              return this._getSegmentCondition();
            case 6:
              segment = _context2.sent;
              jsonQuery = [];
              queryBuilder = new _queryBuilder["default"](this._model, params, this._opts);
              _context2.next = 11;
              return queryBuilder.addFiltersAndJoins(jsonQuery, segment);
            case 11:
              if (!params.search) {
                _context2.next = 15;
                break;
              }
              fieldsSearched = queryBuilder.getFieldsSearched();
              if (!(fieldsSearched.length === 0 && !queryBuilder.hasSmartFieldSearch())) {
                _context2.next = 15;
                break;
              }
              return _context2.abrupt("return", [[], []]);
            case 15:
              if (params.sort) {
                queryBuilder.addSortToQuery(jsonQuery);
              }
              _context2.next = 18;
              return queryBuilder.addProjection(jsonQuery);
            case 18:
              queryBuilder.addSkipAndLimitToQuery(jsonQuery);
              _context2.next = 21;
              return queryBuilder.joinAllReferences(jsonQuery);
            case 21:
              _context2.next = 23;
              return this._model.aggregate(jsonQuery);
            case 23:
              records = _context2.sent;
              return _context2.abrupt("return", [records, fieldsSearched]);
            case 25:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
  }, {
    key: "count",
    value: function () {
      var _count = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3() {
        var params, segment, queryBuilder, jsonQuery, result;
        return _regenerator["default"].wrap(function _callee3$(_context3) {
          while (1) switch (_context3.prev = _context3.next) {
            case 0:
              _context3.next = 2;
              return (0, _scopes["default"])(this._params, this._model, this._user);
            case 2:
              params = _context3.sent;
              _context3.next = 5;
              return this._getSegmentCondition();
            case 5:
              segment = _context3.sent;
              queryBuilder = new _queryBuilder["default"](this._model, params, this._opts);
              _context3.next = 9;
              return queryBuilder.getQueryWithFiltersAndJoins(segment);
            case 9:
              jsonQuery = _context3.sent;
              queryBuilder.addCountToQuery(jsonQuery);
              _context3.next = 13;
              return this._model.aggregate(jsonQuery);
            case 13:
              result = _context3.sent;
              return _context3.abrupt("return", result[0] ? result[0].count : 0);
            case 15:
            case "end":
              return _context3.stop();
          }
        }, _callee3, this);
      }));
      function count() {
        return _count.apply(this, arguments);
      }
      return count;
    }()
  }]);
}();
module.exports = ResourcesGetter;
