// forest-express-mongoose/dist/services/resources-remover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _require = require('./errors'),
  InvalidParameterError = _require.InvalidParameterError;
var ResourcesGetter = require('./resources-getter');
var ResourcesRemover = /*#__PURE__*/function () {
  function ResourcesRemover(model, params, ids, user) {
    (0, _classCallCheck2["default"])(this, ResourcesRemover);
    this._model = model;
    this._params = params;
    this._ids = ids;
    this._user = user;
  }
  return (0, _createClass2["default"])(ResourcesRemover, [{
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              if (!(!Array.isArray(this._ids) || !this._ids.length)) {
                _context.next = 2;
                break;
              }
              throw new InvalidParameterError('`ids` must be a non-empty array.');
            case 2:
              _context.next = 4;
              return this._isAllowed();
            case 4:
              if (!_context.sent) {
                _context.next = 6;
                break;
              }
              return _context.abrupt("return", this._model.deleteMany({
                _id: this._ids
              }));
            case 6:
              return _context.abrupt("return", null);
            case 7:
            case "end":
              return _context.stop();
          }
        }, _callee, this);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
    /**
     * Check if all records maching the provided list of ids are in scope.
     * We can't do that in a single request as mongo does not supports deletes in
     * aggregation pipelines
     */
  }, {
    key: "_isAllowed",
    value: (function () {
      var _isAllowed2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
        var params;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              params = {
                timezone: this._params.timezone,
                filters: JSON.stringify({
                  field: '_id',
                  operator: 'in',
                  value: this._ids
                })
              };
              _context2.t0 = this._ids.length;
              _context2.next = 4;
              return new ResourcesGetter(this._model, null, params, this._user).count();
            case 4:
              _context2.t1 = _context2.sent;
              return _context2.abrupt("return", _context2.t0 === _context2.t1);
            case 6:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function _isAllowed() {
        return _isAllowed2.apply(this, arguments);
      }
      return _isAllowed;
    }())
  }]);
}();
module.exports = ResourcesRemover;
