// forest-express-mongoose/dist/services/resource-updater.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _ = require('lodash');
var Interface = require('forest-express');
var createError = require('http-errors');
var ResourcesGetter = require('./resources-getter');
var utils = require('../utils/schema');
var Flattener = require('./flattener');
var ResourceUpdater = /*#__PURE__*/function () {
  function ResourceUpdater(model, params, record, user) {
    (0, _classCallCheck2["default"])(this, ResourceUpdater);
    this._model = model;
    this._params = params;
    this._record = record;
    this._user = user;
  }
  return (0, _createClass2["default"])(ResourceUpdater, [{
    key: "perform",
    value: function () {
      var _perform = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee() {
        var modelName, schema, recordId, flattenedFields, query;
        return _regenerator["default"].wrap(function _callee$(_context) {
          while (1) switch (_context.prev = _context.next) {
            case 0:
              modelName = utils.getModelName(this._model);
              schema = Interface.Schemas.schemas[utils.getModelName(this._model)];
              recordId = this._record._id;
              _context.next = 5;
              return this._isAllowed(recordId);
            case 5:
              if (_context.sent) {
                _context.next = 7;
                break;
              }
              throw createError(404, "The ".concat(this._model.name, " #").concat(recordId, " does not exist."));
            case 7:
              // NOTICE: Old versions of MongoDB (2.X) seem to refuse the presence of
              //         the _id in the $set. So we remove it. It is useless anyway.
              delete this._record._id;
              flattenedFields = Flattener.getFlattenedFieldsName(schema.fields);
              query = this._model.findByIdAndUpdate(recordId, {
                $set: Flattener.flattenRecordDataForUpdates(this._record, null, flattenedFields)
              }, {
                "new": true,
                runValidators: true
              });
              _.each(schema.fields, function (field) {
                if (field.reference) {
                  query.populate({
                    path: field.field,
                    strictPopulate: false
                  });
                }
              });
              _context.prev = 11;
              return _context.abrupt("return", query.lean().exec());
            case 15:
              _context.prev = 15;
              _context.t0 = _context["catch"](11);
              if (_context.t0.message.indexOf('Cast to') > -1 && _context.t0.message.indexOf('failed for value') > -1) {
                Interface.logger.warn("Cannot update the ".concat(modelName, " #").concat(recordId, " because of a \"type\" key usage (which is a reserved keyword in Mongoose)."));
              } else {
                Interface.logger.error("Cannot update the ".concat(modelName, " #").concat(recordId, " because of an unexpected issue: ").concat(_context.t0));
              }
              throw _context.t0;
            case 19:
            case "end":
              return _context.stop();
          }
        }, _callee, this, [[11, 15]]);
      }));
      function perform() {
        return _perform.apply(this, arguments);
      }
      return perform;
    }()
    /**
     * Check if the record is in scope for the given user.
     * We can't do that in a single request to mongo, as checking scopes requires using
     * an aggregation pipeline to perform $lookups
     */
  }, {
    key: "_isAllowed",
    value: (function () {
      var _isAllowed2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(recordId) {
        var params;
        return _regenerator["default"].wrap(function _callee2$(_context2) {
          while (1) switch (_context2.prev = _context2.next) {
            case 0:
              params = {
                timezone: this._params.timezone,
                filters: JSON.stringify({
                  field: '_id',
                  operator: 'equal',
                  value: recordId
                })
              };
              _context2.next = 3;
              return new ResourcesGetter(this._model, null, params, this._user).count();
            case 3:
              _context2.t0 = _context2.sent;
              return _context2.abrupt("return", _context2.t0 === 1);
            case 5:
            case "end":
              return _context2.stop();
          }
        }, _callee2, this);
      }));
      function _isAllowed(_x) {
        return _isAllowed2.apply(this, arguments);
      }
      return _isAllowed;
    }())
  }]);
}();
module.exports = ResourceUpdater;
