// forest-express-mongoose/dist/services/search-builder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _ = require('lodash');
var Interface = require('forest-express');
var utils = require('../utils/schema');
function SearchBuilder(model, opts, params, searchFields) {
  var _this = this;
  var schema = Interface.Schemas.schemas[utils.getModelName(model)];
  var fieldsSearched = [];
  this.hasSmartFieldSearch = false;
  this.getFieldsSearched = function () {
    return fieldsSearched;
  };
  this.getConditions = /*#__PURE__*/(0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2() {
    var orQuery, pushCondition;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          pushCondition = function _pushCondition(condition, fieldName) {
            orQuery.$or.push(condition);
            fieldsSearched.push(fieldName);
          };
          _this.hasSmartFieldSearch = false;
          orQuery = {
            $or: []
          };
          _.each(model.schema.paths, function (value, key) {
            if (searchFields && !searchFields.includes(value.path) || value.path === model.schema.options.versionKey) {
              return;
            }
            var condition = {};
            var searchValue = params.search.replace(/[-[\]{}()*+!<=:?./\\^$|#\s,]/g, '\\$&');
            var searchRegexp = new RegExp(".*".concat(searchValue, ".*"), 'i');
            if (value.instance === 'ObjectID' || value.instance === 'ObjectId') {
              if (new RegExp('^[0-9a-fA-F]{24}$').test(params.search)) {
                condition[key] = new opts.Mongoose.Types.ObjectId(params.search);
                pushCondition(condition, key);
              }
            } else if (value.instance === 'Number') {
              var searchNumber = Number(params.search);
              if (!Number.isNaN(searchNumber)) {
                condition[key] = searchNumber;
                pushCondition(condition, key);
              }
            } else if (value.instance === 'String') {
              condition[key] = searchRegexp;
              pushCondition(condition, key);
            } else if (value.instance === 'Array') {
              var field = _.find(schema.fields, {
                field: key
              });
              if (_.isArray(field === null || field === void 0 ? void 0 : field.type) && !field.reference) {
                if (field.type[0] === 'String') {
                  condition[key] = searchRegexp;
                  pushCondition(condition, key);
                } else if (field.type[0] === 'Number') {
                  var _searchNumber = Number(params.search);
                  if (!Number.isNaN(_searchNumber)) {
                    condition[key] = _searchNumber;
                    pushCondition(condition, key);
                  }
                } else if (field.type[0].fields && Number.parseInt(params.searchExtended, 10)) {
                  var elemMatch = {
                    $elemMatch: {
                      $or: []
                    }
                  };
                  field.type[0].fields.forEach(function (subField) {
                    var query = {};
                    if (subField.type === 'String' && !value.schema.obj[subField.field].ref) {
                      query[subField.field] = searchRegexp;
                      elemMatch.$elemMatch.$or.push(query);
                    } else if (subField.type === 'Number') {
                      var _searchNumber2 = Number(params.search);
                      if (!Number.isNaN(_searchNumber2)) {
                        query[subField.field] = _searchNumber2;
                        elemMatch.$elemMatch.$or.push(query);
                      }
                    }
                  });
                  condition[key] = elemMatch;
                  pushCondition(condition, key);
                }
              }
            }
          });
          _context2.next = 6;
          return Promise.all(schema.fields.map( /*#__PURE__*/function () {
            var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(field) {
              var condition;
              return _regenerator["default"].wrap(function _callee$(_context) {
                while (1) switch (_context.prev = _context.next) {
                  case 0:
                    if (!field.search) {
                      _context.next = 12;
                      break;
                    }
                    _context.prev = 1;
                    _context.next = 4;
                    return field.search(params.search);
                  case 4:
                    condition = _context.sent;
                    if (condition) {
                      pushCondition(condition, field.field);
                    }
                    _this.hasSmartFieldSearch = true;
                    _context.next = 12;
                    break;
                  case 9:
                    _context.prev = 9;
                    _context.t0 = _context["catch"](1);
                    Interface.logger.error("Cannot search properly on Smart Field ".concat(field.field), _context.t0);
                  case 12:
                  case "end":
                    return _context.stop();
                }
              }, _callee, null, [[1, 9]]);
            }));
            return function (_x) {
              return _ref2.apply(this, arguments);
            };
          }()));
        case 6:
          return _context2.abrupt("return", orQuery.$or.length ? orQuery : {});
        case 7:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  this.getWhere = /*#__PURE__*/function () {
    var _ref3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(jsonQuery) {
      return _regenerator["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            _context3.t0 = jsonQuery;
            _context3.next = 3;
            return _this.getConditions();
          case 3:
            _context3.t1 = _context3.sent;
            _context3.t0.push.call(_context3.t0, _context3.t1);
          case 5:
          case "end":
            return _context3.stop();
        }
      }, _callee3);
    }));
    return function (_x2) {
      return _ref3.apply(this, arguments);
    };
  }();
}
module.exports = SearchBuilder;
