// forest-express-mongoose/dist/adapters/mongoose.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var utils = require('../utils/schema');
var Analyser = require('../utils/field-analyser');

/* eslint-disable */
function unflatten(data) {
  var result = {};
  for (var key in data) {
    var keys = key.split('.');
    keys.reduce(function (result, subKey, index) {
      if (!result[subKey]) {
        if (Number.isNaN(Number(keys[index + 1]))) {
          result[subKey] = keys.length - 1 === index ? data[key] : {};
        } else {
          result[subKey] = [];
        }
      }
      return result[subKey];
    }, result);
  }
  return result;
}
/* eslint-enable */

module.exports = /*#__PURE__*/function () {
  var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(model, opts) {
    var fields, paths, analyser;
    return _regenerator["default"].wrap(function _callee2$(_context2) {
      while (1) switch (_context2.prev = _context2.next) {
        case 0:
          fields = [];
          paths = unflatten(model.schema.paths);
          analyser = new Analyser(model, opts);
          Object.keys(paths).forEach( /*#__PURE__*/function () {
            var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(path) {
              var field;
              return _regenerator["default"].wrap(function _callee$(_context) {
                while (1) switch (_context.prev = _context.next) {
                  case 0:
                    if (!(path === '__v')) {
                      _context.next = 2;
                      break;
                    }
                    return _context.abrupt("return");
                  case 2:
                    field = analyser.getFieldSchema(path, paths[path]);
                    fields.push(field);
                  case 4:
                  case "end":
                    return _context.stop();
                }
              }, _callee);
            }));
            return function (_x3) {
              return _ref2.apply(this, arguments);
            };
          }());
          return _context2.abrupt("return", {
            name: utils.getModelName(model),
            // TODO: Remove nameOld attribute once the lianas versions older than 2.0.0 are minority.
            nameOld: model.collection.name.replace(' ', ''),
            idField: '_id',
            primaryKeys: ['_id'],
            isCompositePrimary: false,
            fields: fields
          });
        case 5:
        case "end":
          return _context2.stop();
      }
    }, _callee2);
  }));
  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}();
